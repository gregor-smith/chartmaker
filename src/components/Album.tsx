import React, { FC } from 'react'
import { css, cx } from 'emotion'
import { Album as AlbumDetails, formatAlbumTitle } from '../state'


type Props = {
    details: AlbumDetails | null
    sizeRem: number
}


const baseStyle = css({
    backgroundColor: 'white',
    margin: '0.15rem'
})


export const Album: FC<Props> = ({ details, sizeRem }) => {
    const style = cx(
        baseStyle,
        css({
            width: `${sizeRem}rem`,
            height: `${sizeRem}rem`
        })
    )
    const image = details === null
        ? null
        : <img src={details.image.smallURL} alt={formatAlbumTitle(details)}/>
    return (
        <div className={style}>
            {image}
        </div>
    )
}
