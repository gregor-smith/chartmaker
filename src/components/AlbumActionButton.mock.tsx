const mock: typeof import('@/components/AlbumActionButton') = {
    AlbumActionButton: props =>
        <div {...props} className='mock-album-action-button'/>
}


export const AlbumActionButton = mock.AlbumActionButton
