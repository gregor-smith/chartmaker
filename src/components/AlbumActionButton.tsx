import { h, FunctionComponent } from 'preact'
import { css } from 'emotion'


type Props = {
    onClick: () => void
    title: string
}


const style = css({
    background: 'none',
    border: 'none'
})


export const AlbumActionButton: FunctionComponent<Props> = props =>
    <button {...props} class={style}/>
