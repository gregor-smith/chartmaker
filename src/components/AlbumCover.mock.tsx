import { cx } from 'emotion'


const mock: typeof import('@/components/AlbumCover') = {
    AlbumCover: ({
        album,
        size,
        overlayClass,
        children,
        highlighted,
        ...props
    }) => {
        const json = JSON.stringify(album)
        const className = cx('mock-album-cover', overlayClass)
        return (
            <div {...props} className={className}>
                {`Album: ${json}`}
                {`Size: ${size}`}
                {`Highlighted: ${highlighted}`}
                {children}
            </div>
        )
    }
}


export const AlbumCover = mock.AlbumCover
