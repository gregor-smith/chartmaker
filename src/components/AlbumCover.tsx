import React, { FC, DragEvent } from 'react'
import { css, cx } from 'emotion'

import { TEXT_COLOUR, ALBUM_PADDING_SIZE, highlightBackgroundStyle } from '@/style'
import { Album } from '@/types'


export type AlbumCoverProps = {
    album: Album
    size: string
    onDragStart: (event: DragEvent<HTMLDivElement>) => void
    onDragOver?: (event: DragEvent<HTMLDivElement>) => void
    onDragEnter?: (event: DragEvent<HTMLDivElement>) => void
    onDrop?: (event: DragEvent<HTMLDivElement>) => void
    onMouseEnter?: () => void
    overlayClass?: string
    highlighted?: boolean
}


const baseContainerStyle = css({
    position: 'relative',
    background: TEXT_COLOUR,
    margin: ALBUM_PADDING_SIZE
})

const imageStyle = css({
    position: 'absolute',
    top: 0,
    left: 0,
    width: '100%',
    maxWidth: '100%',
    height: '100%',
    objectFit: 'contain'
})

// the overlay is given an additional non-emotion class so it can be used as a
// selector in the container's style
const overlaySelector = 'overlay'

const baseOverlayStyle = cx(
    overlaySelector,
    imageStyle,
    css({ zIndex: 10 })
)


export const AlbumCover: FC<AlbumCoverProps> = ({
    album,
    size,
    onDragStart,
    onDragEnter,
    onDragOver,
    onDrop,
    onMouseEnter,
    children,
    overlayClass,
    highlighted
}) => {
    const overlayStyle = cx(baseOverlayStyle, overlayClass)
    const containerStyle = cx(
        baseContainerStyle,
        css({
            width: size,
            height: size,
            [`:not(:hover) .${overlaySelector}`]: {
                display: 'none'
            }
        }),
        highlighted === false
            ? highlightBackgroundStyle
            : undefined
    )

    let image: JSX.Element | undefined
    if (!album.placeholder) {
        image = (
            // lazy loading causes problems with html2canvas so keep it default
            <img className={imageStyle}
                src={album.url}
                alt={album.name}
                crossOrigin='anonymous'/>
        )
    }

    return (
        <div className={containerStyle}
                draggable={!album.placeholder}
                onDragStart={onDragStart}
                onDragEnter={onDragEnter}
                onDragOver={onDragOver}
                onDrop={onDrop}
                onMouseEnter={onMouseEnter}
                title={album.placeholder ? undefined : album.name}>
            {image}
            <div className={overlayStyle}>
                {children}
            </div>
        </div>
    )
}
