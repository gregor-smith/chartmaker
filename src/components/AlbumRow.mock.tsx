const mock: typeof import('@/components/AlbumRow') = {
    AlbumRow: ({ children }) =>
        <div className='mock-album-row'>
            {children}
        </div>
}


export const AlbumRow = mock.AlbumRow
