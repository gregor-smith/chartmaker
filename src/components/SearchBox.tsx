import React, { FC } from 'react'
import { css } from 'emotion'

import { DispatchProps } from '@/reducer'
import { SIDEBAR_LABEL_PADDING_SIZE, ERROR_TEXT_COLOUR } from '@/style'
import { SearchState } from '@/types'
import { Label } from '@/components/Label'
import { ControlledForm } from '@/components/ControlledForm'
import { SidebarGroup } from '@/components/SidebarGroup'
import { ControlledInput } from '@/components/ControlledInput'


const id = 'search'


type Props = DispatchProps<
    | 'UpdateSearchQuery'
    | 'SendSearchRequest'
    | 'CancelSearchRequest'
> & {
    searchState: SearchState
}


const errorStyle = css({
    color: ERROR_TEXT_COLOUR,
    marginBottom: SIDEBAR_LABEL_PADDING_SIZE,
    userSelect: 'none'
})


export const SearchBox: FC<Props> = ({ dispatch, searchState }) => {
    function sendSearchRequest() {
        dispatch({ tag: 'CancelSearchRequest' })
        dispatch({ tag: 'SendSearchRequest' })
    }

    function updateQuery(query: string) {
        dispatch({
            tag: 'UpdateSearchQuery',
            query
        })
    }

    const errorMessage = searchState.tag === 'Error'
        ? (
            <div className={errorStyle}>
                {searchState.message}
            </div>
        )
        : null

    return (
        <SidebarGroup>
            <ControlledForm onSubmit={sendSearchRequest}>
                <Label target={id}>
                    Search for albums
                </Label>
                {errorMessage}
                <ControlledInput id={id}
                    value={searchState.query}
                    onChange={updateQuery}
                    disabled={searchState.tag === 'Loading'}/>
            </ControlledForm>
        </SidebarGroup>
    )
}
