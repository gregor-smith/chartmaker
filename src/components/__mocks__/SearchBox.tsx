import React, { FC } from 'react'

import { SearchBoxProps } from '@/components/SearchBox'


export const SearchBox: FC<SearchBoxProps> = ({
    searchState
}) => {
    const json = JSON.stringify(searchState)
    return (
        <div className='mock-search-box'>
            {`Search state: ${json}`}
        </div>
    )
}
