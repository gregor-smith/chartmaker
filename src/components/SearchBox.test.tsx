import { render } from 'react-dom'
import { act } from 'react-dom/test-utils'

import type { SearchState } from '@/types'
import { SearchBox, id } from '@/components/SearchBox'
import type { Action } from '@/reducer'

import { RenderContainer, ignore, fireEvent } from '../utils'


// no need to mock the ControlledForm because it has no styles or anything,
// it's just a plain form once rendered
jest.mock('@/components/SidebarGroup')
jest.mock('@/components/Label')
jest.mock('@/components/ControlledInput')


const container = new RenderContainer()


test.each<SearchState>([
    {
        tag: 'Waiting',
        query: 'Test query'
    },
    {
        tag: 'Loading',
        query: 'Test query',
        // only the reducer touches this field so it's ok
        controller: undefined as any
    },
    {
        tag: 'Error',
        query: 'Test query',
        message: 'Test error message'
    },
    {
        tag: 'Complete',
        query: 'Test query',
        // component does not touch the albums at all so no need to specify any
        albums: []
    }
])('renders form with labelled text input', searchState => {
    render(
        <SearchBox dispatch={ignore}
            searchState={searchState}/>,
        container.element
    )

    expect(container.element).toMatchSnapshot()
})


test('dispatches action on input change', () => {
    const mock = jest.fn<void, [ Action ]>()

    render(
        <SearchBox dispatch={mock}
            searchState={{ tag: 'Waiting', query: 'Test query' }}/>,
        container.element
    )

    act(() =>
        fireEvent(
            'change',
            container.element?.querySelector(`#${id}`),
            { target: { value: 'New query' } }
        )
    )

    expect(mock).toHaveBeenCalledTimes(1)
    expect(mock).toHaveBeenCalledWith<[ Action ]>({
        tag: 'UpdateSearchQuery',
        query: 'New query'
    })
})


test('dispatches action on form submit', () => {
    const mock = jest.fn<void, [ Action ]>()

    render(
        <SearchBox dispatch={mock}
            searchState={{ tag: 'Waiting', query: 'Test query' }}/>,
        container.element
    )

    act(() =>
        fireEvent(
            'submit',
            container.element?.querySelector(`form`)
        )
    )

    expect(mock).toHaveBeenCalledTimes(2)
    expect(mock).toHaveBeenNthCalledWith<[ Action ]>(
        1,
        { tag: 'CancelSearchRequest' }
    )
    expect(mock).toHaveBeenNthCalledWith<[ Action ]>(
        2,
        { tag: 'SendSearchRequest' }
    )
})
