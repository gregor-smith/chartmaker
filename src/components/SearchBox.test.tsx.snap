// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders form with labelled text input 1`] = `
<div
  id="test-render-container"
>
  <div
    class="mock-sidebar-group"
  >
    <form>
      <div
        class="mock-label"
      >
        Target: search
        Search for albums
      </div>
      <div
        class="mock-controlled-input"
        id="search"
      >
        Type: search
        Value: Test query
        Disabled: false
      </div>
    </form>
  </div>
</div>
`;

exports[`renders form with labelled text input 2`] = `
<div
  id="test-render-container"
>
  <div
    class="mock-sidebar-group"
  >
    <form>
      <div
        class="mock-label"
      >
        Target: search
        Search for albums
      </div>
      <div
        class="mock-controlled-input"
        id="search"
      >
        Type: search
        Value: Test query
        Disabled: true
      </div>
    </form>
  </div>
</div>
`;

exports[`renders form with labelled text input 3`] = `
.emotion-0 {
  color: red;
  margin-bottom: 0.25rem;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

<div
  id="test-render-container"
>
  <div
    class="mock-sidebar-group"
  >
    <form>
      <div
        class="mock-label"
      >
        Target: search
        Search for albums
      </div>
      <div
        class="emotion-0"
      >
        Test error message
      </div>
      <div
        class="mock-controlled-input"
        id="search"
      >
        Type: search
        Value: Test query
        Disabled: false
      </div>
    </form>
  </div>
</div>
`;

exports[`renders form with labelled text input 4`] = `
<div
  id="test-render-container"
>
  <div
    class="mock-sidebar-group"
  >
    <form>
      <div
        class="mock-label"
      >
        Target: search
        Search for albums
      </div>
      <div
        class="mock-controlled-input"
        id="search"
      >
        Type: search
        Value: Test query
        Disabled: false
      </div>
    </form>
  </div>
</div>
`;
